presto-orc/src/main/java/com/facebook/presto/orc/DictionaryCompressionOptimizer.java

```diff
@@ -14,54 +14,90 @@
+ * so for comparable column sizes, direct encoding is mostly better.
+ * <p>
+ * Note: Dictionary writer might use more memory as they over-allocate dictionary sizes as the writers
+ * build dictionary as they see new data. The hash tables implementation in the dictionary writer's allocate
+ * hash buckets in power of 2. So after a million entries, the overallocation consumes large amount of memory.
+ * <p>
+ * DictionaryCompressionOptimizer functionality can be controlled by the following configs to the constructor.
+ * <p>
+ * 1. dictionaryMemoryMaxBytes -> Max size of the dictionary when all columns are read. Note: Writer
+ * might consume more memory due to the over-allocation.
+ * <p>
+ * 2. dictionaryMemoryAlmostFullRangeBytes -> When the dictionary size, exceeds dictionaryMaxMemoryBytes
+ * dictionary columns will be converted to direct to reduce the dictionary size. By setting a range
+ * the stripe can be flushed, before the dictionary is full. When dictionary size is higher than
+ * (dictionaryMemoryMaxBytes - dictionaryMemoryAlmostFullRangeBytes), it is considered almost full
+ * and is ready for flushing. This setting is defined as a delta on dictionaryMemoryMaxBytes for backward compatibility.
+ * <p>
+ * 3. dictionaryUsefulCheckColumnSizeBytes -> Columns start with dictionary encoding and when the dictionary memory
+ * is almost full, usefulness of the dictionary is measured. For large dictionaries (> 40 MB) the check
+ * might happen very late and large dictionary might cause writer to OOM due to writer over allocating for
+ * dictionary growth. When a dictionary for a column grows beyond the dictionaryUsefulCheckColumnSizeBytes the
+ * dictionary usefulness check will be performed and if dictionary is not useful, it will be converted to direct.
+ * <p>
+ * 4. dictionaryUsefulCheckPerChunkFrequency -> dictionaryUsefulCheck could be costly if performed on every chunk.
+ * The dictionaryUsefulCheck will be performed when a column dictionary is above the dictionaryUsefulCheckColumnSizeBytes
+ * and per every dictionaryUsefulCheckPerChunkFrequency chunks written.
+ */
 public class DictionaryCompressionOptimizer
 {
     private static final double DICTIONARY_MIN_COMPRESSION_RATIO = 1.25;
 
-    // Instead of waiting for the dictionary to fill completely, which would force a column into
-    // direct mode, close the stripe early assuming it has hit the minimum row count.
-    static final DataSize DICTIONARY_MEMORY_MAX_RANGE = new DataSize(4, Unit.MEGABYTE);
-
-    static final DataSize DIRECT_COLUMN_SIZE_RANGE = new DataSize(4, Unit.MEGABYTE);
+    static final DataSize DIRECT_COLUMN_SIZE_RANGE = new DataSize(4, MEGABYTE);
 
-    private final Set<DictionaryColumnManager> allWriters;
-    private final Set<DictionaryColumnManager> directConversionCandidates = new HashSet<>();
+    private final List<DictionaryColumnManager> allWriters;
+    private final List<DictionaryColumnManager> directConversionCandidates = new ArrayList<>();
```

Comment:
There is quite a lot of removal magic happening around this list. Should we consider using a LinkedList?

Reply:
I wanted to use LinkedList, but there is no way to capture the reference to a middle element of the list and remove it later.  So I have to go for the ArrayList and do swap the element with last to preserve O(1) behavior.